### PR TITLE
refactor(scheduler): optimize performance. for batch 1, skip table qu…

### PR DESCRIPTION
**Pull request by Ryan-Git**
…eue; for batches, use ack callback instead of workingset.

result: (ns/10w msg)
BenchmarkStream/10worker-32table-1batch 406520166 -> 354436043 (12%)
BenchmarkStream/10worker-32table-10batch 579279148 -> 466868294 (19%)
BenchmarkBatch/10worker-32table-1batch  550858716 -> 429352945 (22%)
BenchmarkBatch/10worker-32table-10batch  682855587 -> 464386811 (32%)

### CODE REVIEW

**Comment by Ryan-Git** [comment]
this closes #68

**Comment by Ryan-Git** [comment]
PLAT @ming-relax

File: pkg/schedulers/batch_table_scheduler/batch_table_scheduler.go [hunk 1]
```diff
@@ -324,90 +333,136 @@ func (scheduler *batchScheduler) Close() {
...
+
+			case h, ok := <-latchC:
+				if !ok {
+					scheduler.latchWg.Done()
+					return
+				}
+				latch[h]--
+				if latch[h] == 0 {
+					delete(latch, h)
+					if len(batch) > 0 && len(latchC) == 0 {
```
Comment (ming535) [inline pkg/schedulers/batch_table_scheduler/batch_table_scheduler.go]: Why do we need to make sure `len(latchC) == 0`? @Ryan-Git

Comment (Ryan-Git) [reply pkg/schedulers/batch_table_scheduler/batch_table_scheduler.go]: it's just an optimization. if there're still locks hold, the next batch is very likely to fail. And for each new coming msg, the batch will be retried.

**Comment by ming535** [comment]
@Ryan-Git PTAL

File: pkg/core/output.go [hunk 2]
```diff
@@ -1,7 +1,7 @@
 package core
 
 type Output interface {
-	Execute(msgs []*Msg) error
+	Execute(workerIndex int, msgs []*Msg) error
```
Comment (Ryan-Git) [inline pkg/core/output.go]: why need worker index?

File: pkg/inputs/mongostream/oplog_tailer.go [hunk 3]
```diff
@@ -209,7 +211,12 @@ func (tailer *OplogTailer) Run() {
 			}
 			c.Add(1)
 			msg.InputStreamKey = utils.NewStringPtr("mongooplog")
-			msg.OutputStreamKey = utils.NewStringPtr(outputStreamKey(msg.Oplog))
+			idKey := outputStreamKey(msg.Oplog)
```
Comment (Ryan-Git) [inline pkg/inputs/mongostream/oplog_tailer.go:214]: mongo also has unique index...

File: pkg/inputs/tidb_kafka/binlog_tailer.go [hunk 4]
```diff
@@ -268,7 +270,12 @@ func deserialize(raw *pb.Column, colType string) interface{} {
 
 func (t *BinlogTailer) dispatchMsg(msg *core.Msg) error {
 	msg.InputStreamKey = utils.NewStringPtr("tidbbinlog")
-	msg.OutputStreamKey = utils.NewStringPtr(msg.GetPkSign())
+	pkSign := msg.GetPkSign()
+	h, err := hashstructure.Hash(pkSign, nil)
```
Comment (Ryan-Git) [inline pkg/inputs/tidb_kafka/binlog_tailer.go:274]: not support uk?

File: pkg/outputs/async_kafka/async_kafka.go [hunk 5]
```diff
@@ -192,7 +188,8 @@ func (output *AsyncKafka) Execute(msgs []*core.Msg) error {
 		if err != nil {
 			return errors.Annotatef(err, "topic: %v", topic)
 		}
-		partition := int(utils.GenHashKey(*msg.OutputStreamKey)) % len(partitions)
+
+		partition := workerIndex % len(partitions)
```
Comment (Ryan-Git) [inline pkg/outputs/async_kafka/async_kafka.go]: this does not help. maybe we can keep using primary key.

**Comment by ming535** [comment]
added an issue and update doc https://github.com/moiot/gravity/issues/128

File: pkg/outputs/async_kafka/async_kafka.go [hunk 6]
```diff
@@ -189,7 +189,13 @@ func (output *AsyncKafka) Execute(workerIndex int, msgs []*core.Msg) error {
 			return errors.Annotatef(err, "topic: %v", topic)
 		}
 
-		partition := workerIndex % len(partitions)
+		// data with the same primary key goes to the same partition,
+		// if there is no primary key, use unique index;
+		// if there is no primary key and unique index, use partition 0.
```
Comment (Ryan-Git) [inline pkg/outputs/async_kafka/async_kafka.go]: if there is no primary key and unique index, use random partition should be used.

Comment (Ryan-Git) [inline pkg/outputs/async_kafka/async_kafka.go]: if there is no primary key and unique index, random partition should be used.

Comment (ming535) [reply pkg/outputs/async_kafka/async_kafka.go]: changed @Ryan-Git

**Review by ming535 — APPROVED** [review]

**Pull request merged**